client/pyscripts/fm11rf08_full.py

```diff
@@ -0,0 +1,942 @@
+
+	return True
+
+#+=============================================================================
+# Run key recovery script
+# >> "keyfile"
+#==============================================================================
+def recoverKeys():
+	global keyfile
+
+	badrk   = 0     # 'bad recovered key' count (ie. not recovered)
+
+	lprint(prompt)
+	lprint(f"{prompt} Running recovery script, ETA: Less than 30 minutes")
+
+	lprint(prompt)
+	lprint(f'{prompt} `-._,-\'"`-._,-"`-._,-\'"`-._,-\'"`-._,-\'"`-._,-\'"`-._,-\'"`-._,-\'"`-._,')
+
+	r = recovery(quiet=False)
+	keyfile = r['keyfile']
+	rkey    = r['found_keys']
+	fdump   = r['dumpfile']
+	rdata   = r['data']
+
+	lprint(f'{prompt} `-._,-\'"`-._,-"`-._,-\'"`-._,-\'"`-._,-\'"`-._,-\'"`-._,-\'"`-._,-\'"`-._,')
+
+	for k in range(0, 16+1):
+		for ab in [0, 1]:
+			if rkey[k][ab] == "":
+				if badrk == 0:  lprint(f"{prompt} Some keys were not recovered: ", end='')
+				else:           lprint(f", ", end='')
+				badrk += 1
+
+				kn = k
+				if kn > 15:  kn += 16
+				lprint(f"[{kn}/", end='')
+				lprint("A]" if ab == 0 else "B]", end='')
+	if badrk > 0:  lprint("")
+
+#+=============================================================================
+# Verify keys
+# >> "key[][]"
+# >> mad!
+#==============================================================================
+def verifyKeys():
+	global  key
+	global  mad
+
+	badk = 0
+	mad  = False
+
+	lprint(f"{prompt} Check keys..")
+
+	for sec in range (0,16+1):  # 16 normal, 1 dark
+		sn = sec
+		if (sn > 15):  sn = sn + 16
+
+		for ab in [0, 1]:
+			bn  = (sec * 4) + 3
+			if bn >= 64:  bn += 64
+
+			cmd = f"hf mf rdbl -c {ab} --key {key[sec][ab].hex()} --blk {bn}"
+			lprint(f"{prompt}   `{cmd}`", end='', flush=True)
+
+			res = p.console(f"{cmd}")
+			lprint(" " * (3-len(str(bn))), end="")
+			if res == 0:
+				lprint(" ... PASS", end="")
+			else:
+				lprint(" ... FAIL", end="")
+				badk += 1
+				key[sec][ab] = ""
+
+			# check for Mifare Application Directory
+			if (sec == 0) and (ab == 0) \
+			   and (key[0][0] == b'\xa0\xa1\xa2\xa3\xa4\xa5'):
+				mad = True
+				lprint(" - MAD Key")
+			else:
+				lprint("")
+
+	# We need to flush all the output we just collected <shrug>
+	_ = p.grabbed_output
+
+	if badk > 0:
+		lprint(f"{prompt} ! {badk} bad key", end='')
+		lprint("s exist" if badk != 1 else " exists")
+		rv = False
+
+	else:
+		lprint(f"{prompt} All keys verified OK")
+		rv = True
+
+	if mad == True:
+		lprint(f"{prompt} MAD key detected")
+
+	return rv
+
+#+=============================================================================
+# Read all block data - INCLUDING Dark blocks
+# >> blkn
+# >> "data[]"
+'''
+[=]   # | sector 00 / 0x00                                | ascii
+[=] ----+-------------------------------------------------+-----------------
+[=]   0 | 5C B4 9C A6 D2 08 04 00 04 59 92 25 BF 5F 70 90 | \........Y.%._p.
+'''
+#==============================================================================
+def readBlocks():
```

Comment:
redundant with the dump file already provided by recovery()
or call it verifyDump as you have the verifyKeys...

Reply:
Considerations:
~ This function reads the 8 Dark blocks in addition to the 64 "normal" blocks (which were read by `recovery()`)
~ `verifyKeys` takes the known keys and verifies that they are still valid. ...because we would like to avoid waiting "<30mins" if possible
~ `readBlocks` deliberately re-reads the card each time it is run, so as to detect edits to the card ...it only takes a few seconds ...I have to read it all to verify nothing has changed, at which point, well, I've just read it all ...There may well be credit to writing a `verifyBlocks()` function to detect changes - but that would need to call `readblock()` and `loadOldFile()`
~ The feature bloat had already crept in when I started detecting and decoding various tags - I just had to draw the line somewhere (for now)

Reply:
if you need to read the card, using "hf mf ecfill -c 4 -k ..." is wayyy faster.
Then, yes, for now, you've to dump manually the 8 avblocks because as said above, it is not yet supported in ecfill or any other mfc emulation command. (I think even letting a "hf mf ecfill --4k -c 4 -k..." dumping and failing over unexisting sectors is still faster)